File: Anole/HTMLTableElement.vb
Namespace Anole

    Public Class HTMLTableElement : Inherits HTMLElement

        Protected myTargetWidth As Integer

        Public Property TargetWidth() As Integer
            Get
                Return myTargetWidth
            End Get
            Set(ByVal value As Integer)
                myTargetWidth = value
            End Set
        End Property

        Protected myBorderSize As Integer

        Public Property BorderSize() As Integer
            Get
                Return myBorderSize
            End Get
            Set(ByVal value As Integer)
                myBorderSize = value
            End Set
        End Property

        Protected myCellSpacing As Integer

        Public Property CellSpacing() As Integer
            Get
                Return myCellSpacing
            End Get
            Set(ByVal value As Integer)
                myCellSpacing = value
            End Set
        End Property

        Protected myCellPadding As Integer

        Public Property CellPadding() As Integer
            Get
                Return myCellPadding
            End Get
            Set(ByVal value As Integer)
                myCellPadding = value
            End Set
        End Property

        Public Sub New(ByRef document As HTMLDocument, ByVal parent As HTMLElement, ByVal xnode As XmlNode)
            Initialize(document, parent, xnode)

            'Get backcolor or use document default color
            Me.BackColor = Me.ParseColor(xnode, "bgcolor", document.BackColor)
            Me.myTargetWidth = Me.ParseInt(xnode, "width", 0)
            Me.myBorderSize = Me.ParseInt(xnode, "borders", 1)
            Me.myCellPadding = Me.ParseInt(xnode, "cellpadding", 0)
            Me.myCellSpacing = Me.ParseInt(xnode, "cellspacing", 0)
            ParseChildren(xnode)
        End Sub

        Public Overloads Overrides Sub Layout(ByVal renderer As HTMLRenderer, ByVal g As System.Drawing.Graphics)
            'Determine number of rows and columns
            Dim nRowCount As Integer = 0
            Dim nColCount As Integer = 0

            nRowCount = Me.Children.Count
            Dim firstrow As HTMLTRElement = TryCast(Me.Children(0), HTMLTRElement)

            If nRowCount > 0 Then
                nColCount = firstrow.Children.Count
            Else
                System.Windows.Forms.MessageBox.Show("No rows in table")
                Return
            End If

            'If the table width is not set, assume
            'that the width is the width of the renderer
            If myTargetWidth = 0 Then myTargetWidth = renderer.TargetWidth

            'Determine an initial column count
            Dim nTargetColumnWidth As Integer = myTargetWidth \ nColCount

            'Create an array of column widths and
            'set to the intial column width
            Dim nColumnWidths As Integer() = New Integer(nColCount - 1) {}
            For i As Integer = 0 To nColCount - 1
                nColumnWidths(i) = nTargetColumnWidth
            Next i

            'TBD: loop thru the first tr and see if any
            'td elements override the default column width

            'Create an array to track row-heights
            Dim nRowHeights As Integer() = New Integer(nRowCount - 1) {}

            'Give each row the column widths
            Dim trelement As HTMLTRElement

            For i As Integer = 0 To Me.Children.Count - 1
                trelement = TryCast(Children(i), HTMLTRElement)
                nRowHeights(i) = trelement.ComputeRowDimensions(nColumnWidths, g)
            Next i

            'Compute table height
            Dim tableheight As Integer = 0

            For i As Integer = 0 To nRowHeights.Length - 1
                nRowHeights(i) = 10
                tableheight += nRowHeights(i)
            Next

            'Compute table width
            Dim tablewidth As Integer = 0
            For i As Integer = 0 To nColumnWidths.Length - 1
                tablewidth += nColumnWidths(i)
            Next i

            'Start a new band
            Dim band As HTMLRenderBand = renderer.NewBand()

            band.AddItem(New HTMLRenderTableItem(nRowCount, nColCount, tablewidth, tableheight, Me, nColumnWidths, _
             nRowHeights))

            'Add a final band
            renderer.NewBand()
        End Sub

    End Class

    Public Class HTMLTRElement : Inherits HTMLElement

        Protected myColumnWidths As Integer()

        Public Sub New(ByRef document As HTMLDocument, ByVal parent As HTMLElement, ByVal xnode As XmlNode)
            Initialize(document, parent, xnode)
            Me.BackColor = Me.ParseColor(xnode, "bgcolor", parent.BackColor)
            ParseChildren(xnode)
        End Sub

        ' Given an array of column widths fills in the max of columnwidths
        ' Returns the height of the row
        Public Function ComputeRowDimensions(ByRef nColumnWidths As Integer(), ByVal g As Graphics) As Integer
            Dim r As HTMLRenderer
            Dim td As HTMLTDElement
            Dim rowheight As Integer = 0

            For i As Integer = 0 To Me.Children.Count - 1
                'Create a renderer and have it render our td item and all its children
                td = TryCast(Me.Children(i), HTMLTDElement)
                r = New HTMLRenderer(td, nColumnWidths(i))

                'Store the max of either the computed width or the original width
                nColumnWidths(i) = Math.Max(nColumnWidths(i), r.Width)

                'Store the height of row
                rowheight = Math.Max(rowheight, r.Height)
            Next i

            Return rowheight
        End Function

    End Class

    Public Class HTMLTDElement : Inherits HTMLElement

        Protected myTargetWidth As Integer

        Public Property TargetWidth() As Integer
            Get
                Return myTargetWidth
            End Get
            Set(ByVal value As Integer)
                myTargetWidth = value
            End Set
        End Property

        Public Sub New(ByRef document As HTMLDocument, ByVal parent As HTMLElement, ByVal xnode As XmlNode)
            Initialize(document, parent, xnode)
            Me.BackColor = Me.ParseColor(xnode, "bgcolor", parent.BackColor)
            Me.Alignment = Me.ParseAlignment(xnode, HTMLAlignment.Left)
            Me.myTargetWidth = Me.ParseInt(xnode, "width", 0)

            ParseChildren(xnode)
        End Sub

    End Class

End Namespace

File: Anole/HTMLTableElement.cs
using System;
using System.Xml;
using System.Drawing;

namespace Anole
{
	/// <summary>
	/// Summary description for HTMLTableElement.
	/// </summary>
	public class HTMLTableElement : HTMLElement
	{
		protected int myTargetWidth;
		public int TargetWidth
		{
			get{return myTargetWidth;}
			set{myTargetWidth=value;}
		}

		protected int myBorderSize;
		public int BorderSize
		{
			get{return myBorderSize;}
			set{myBorderSize=value;}
		}

		protected int myCellSpacing;
		public int CellSpacing
		{
			get{return myCellSpacing;}
			set{myCellSpacing=value;}
		}

		protected int myCellPadding;
		public int CellPadding
		{
			get{return myCellPadding;}
			set{myCellPadding=value;}
		}

		public HTMLTableElement(ref HTMLDocument document, HTMLElement parent, XmlNode xnode)
		{
			Initialize(ref document, parent, xnode);

			//Get backcolor or use document default color
			this.BackColor=this.ParseColor(xnode,"bgcolor",document.BackColor);
			this.myTargetWidth=this.ParseInt(xnode,"width",0);
			this.myBorderSize=this.ParseInt(xnode,"borders",1);
			this.myCellPadding=this.ParseInt(xnode,"cellpadding",0);
			this.myCellSpacing=this.ParseInt(xnode,"cellspacing",0);
			ParseChildren(xnode);
		}

		public override void Layout(HTMLRenderer renderer, System.Drawing.Graphics g)
		{
			//Determine number of rows and columns
			int nRowCount=0;
			int nColCount=0;
			nRowCount=this.Children.Count;
			HTMLTRElement firstrow=this.Children[0] as HTMLTRElement;
			if(nRowCount>0)
			{
				nColCount=firstrow.Children.Count;
			}
			else
			{
				System.Windows.Forms.MessageBox.Show("No rows in table");
				return;
			}

			//If the table width is not set, assume
			//that the width is the width of the renderer
			if(myTargetWidth==0)
			{
				myTargetWidth=renderer.TargetWidth;
			}

			//Determine an initial column count
			int nTargetColumnWidth=myTargetWidth/nColCount;

			//Create an array of column widths and
			//set to the intial column width
			int[] nColumnWidths=new int[nColCount];
			for(int i=0; i<nColCount; i++)
			{
				nColumnWidths[i]=nTargetColumnWidth;
			}
			//TBD: loop thru the first tr and see if any
			//td elements override the default column width

			//Create an array to track row-heights
			int[] nRowHeights=new int[nRowCount];


			//Give each row the column widths
			HTMLTRElement trelement;
			for(int i=0; i<this.Children.Count; i++)
			{
				trelement=Children[i] as HTMLTRElement;
				nRowHeights[i]=trelement.ComputeRowDimensions(ref nColumnWidths, g);
				
			}

			//Compute table height
			int tableheight=0;
			for(int i=0; i<nRowHeights.Length; i++)
			{
				nRowHeights[i]=10;
				tableheight+=nRowHeights[i];
			}

			//Compute table width
			int tablewidth=0;
			for(int i=0; i<nColumnWidths.Length; i++)
			{
				tablewidth+=nColumnWidths[i];
			}


			//Start a new band
			HTMLRenderBand band=renderer.NewBand();

			band.AddItem(new HTMLRenderTableItem(nRowCount,nColCount,
				tablewidth,tableheight,this,nColumnWidths,nRowHeights));

			//Add a final band
			renderer.NewBand();

		}

	}

	public class HTMLTRElement : HTMLElement
	{
		protected int[] myColumnWidths;

		public HTMLTRElement(ref HTMLDocument document, HTMLElement parent, XmlNode xnode)
		{
			Initialize(ref document, parent, xnode);
			this.BackColor=this.ParseColor(xnode,"bgcolor",parent.BackColor);
			ParseChildren(xnode);
		}


		/// <summary>
		/// Given an array of column widhts fills in the max of columnwidths
		/// Returns the height of the row.
		/// </summary>
		/// <param name="nColumnWidths"></param>
		/// <param name="g"></param>
		/// <returns></returns>
		public int ComputeRowDimensions(ref int[] nColumnWidths, Graphics g)
		{
			HTMLRenderer r;
			HTMLTDElement td;
			int rowheight=0;
			for(int i=0; i<this.Children.Count; i++)
			{
				//Create a renderer and have it render
				//our td item and all it's children
				td=this.Children[i] as HTMLTDElement;
				r=new HTMLRenderer(td,nColumnWidths[i]);

				//Store the max of either the computed with
				//or the original width

				nColumnWidths[i]=Math.Max(nColumnWidths[i],r.Width);
				
				//Store the height of row
				rowheight=Math.Max(rowheight,r.Height);

			}
			return rowheight;
		}

		

	}

	public class HTMLTDElement : HTMLElement
	{
		protected int myTargetWidth;
		public int TargetWidth
		{
			get{return myTargetWidth;}
			set{myTargetWidth=value;}
		}

		public HTMLTDElement(ref HTMLDocument document, HTMLElement parent, XmlNode xnode)
		{
			Initialize(ref document, parent, xnode);
			this.BackColor=this.ParseColor(xnode,"bgcolor",parent.BackColor);
			this.Alignment=this.ParseAlignment(xnode,HTMLAlignment.Left);
			this.myTargetWidth=this.ParseInt(xnode,"width",0);

			ParseChildren(xnode);
		}
	}
}
